tutorials/pubsub-spring-dedup-messages/index.md

```diff
@@ -0,0 +1,263 @@
+  // name `receiveDedupedMessagesFromDataflow-in-0` in application.properties.
+  @Bean
+  Consumer<Message<String>> receiveDedupedMessagesFromDataflow() {
+    return msg -> {
+      System.out.println("\tDE-DUPED message: \"" + msg.getPayload() + "\".");
+    }
+```
+
+### Specify a Cloud Pub/Sub topic and subscription for the sink
+Spring can find the Cloud Pub/Sub subscription that your code should receive data from when you provide a subscription name for your sink in `application.properties` using an input binder. Here, the input binder's binding name is `receiveDedupedMessagesFromDataflow-in-0`. Because a subscription cannot exist without a topic, a topic must also be specified. The input binder's destination is your Pub/Sub topic and its consumer group is your Pub/Sub subscription. Note that only input bindings have consumer groups. For more information, see [Common Binding Properties].
+
+In [application.properties](pubsub-spring/src/main/resources/application.properties):
+
+[embedmd]:# (pubsub-spring/src/main/resources/application.properties /.*Data coming/ /=subscriptionFromDataflow/)
+```properties
+# Data coming from Cloud Pub/Sub to a Spring Cloud Stream sink defined in the
+# Consumer bean `receiveDedupedMessagesFromDataflow`. The application
+# will create the Cloud Pub/Sub topic `topicFromDataflow` and subscription
+# `topicFromDataflow.subscriptionFromDataflow` if they do not exist.
+spring.cloud.stream.bindings.receiveDedupedMessagesFromDataflow-in-0.destination=topicFromDataflow
+spring.cloud.stream.bindings.receiveDedupedMessagesFromDataflow-in-0.group=subscriptionFromDataflow
+```
+
+### Construct a Pub/Sub message with custom attributes
+
+Cloud Pub/Sub lets you publish messages with custom attributes, which are key value pairs. To construct such a message in your Spring Boot application, you can specify the attributes as headers. 
+
+In [DataEntryController.java](pubsub-spring/src/main/java/com/google/example/DataEntryController.java):
+
+[embedmd]:# (pubsub-spring/src/main/java/com/google/example/DataEntryController.java java /  \/\/ Headers/ /;/)
+```java
+  // Headers become Pub/Sub message attributes.
+    Message<String> message = MessageBuilder
+      .withPayload(data)
+      .setHeader("key", key)
+      .build();
+```
+
+### Run the application
+To start your application, navigate to `pubsub-spring/` and run: 
+
+```shell script
+mvn spring-boot:run
+```
+
+Check that your app has started successfully by pointing your browser to `localhost:8080`. You should be able to send messages using the form there. At this point, Spring will have automatically created the Cloud Pub/Sub topic that you have specified in `application.properties` to publish your message to. You can view Publish Message Request Count and Publish Message Operation Count in the topic details in [Cloud Console for Pub/Sub Topic] to verify that publishing to Cloud Pub/Sub is successful.  
```

Comment:
This paragraph is hard to follow and may be easier to follow as a list of steps the reader can do to verify that the app is working.